### PR TITLE
 Enhance/document msg listener heuristic

### LINKED ISSUE #3135: Allow opt-out for MessageListener matcher performance optimization
#2930 added a performance optimization by pre-filtering instrumentation candidates for our `MessageListener` instrumentation.
Unfortunately, there is currently no way of restoring the old behaviour: It is not possible to make the agent scan all packages for `MessageListeners`. If a user has no knowledge of where the `MessageListeners` are placed, he has no chance of instrumenting them.

A simple, backwards-compatible fix would be to enhance the `elastic.apm.jms_listener_packages` config option to accept wildcard matchers instead of strings. This way, a user can set `jms_listener_packages=*` to make the agent scan all packages.

A downside of this approach is a bit of inconsistency in the configuration: Other wildcard options such as `classes_excluded_from_instrumentation` match the wildcards against the entire class name, whereas `jms_listener_packages` only checks that the classname begins with the provided pattern / string.

### PR DESCRIPTION
**Pull request by SylvainJuge**
## What does this PR do?

Fixes #3135 
Relates to #3259 

I had to move the test JMS classes to a package outside of `co.elastic.apm` as `co.elastic.apm` is already covered by the test configuration which is automatically loaded from classpath.

## Checklist

- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] Added an API method or config option? Document in which version this will be introduced
  - [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: docs/configuration.asciidoc [hunk 1]
```diff
@@ -2430,9 +2430,12 @@ Prepending an element with `(?-i)` makes the matching case sensitive.
 ==== `jms_listener_packages` (performance added[1.36.0])
 
 Defines which packages contain JMS MessageListener implementations for instrumentation.
-When set to a non-empty value, only the classes matching configuration will be instrumented.
+When empty (default), only the inner-classes or that have 'Listener' or 'Message' in their names are considered.
```
Comment (jackshirazi) [inline docs/configuration.asciidoc:2433]: only which inner classes? any in the whole application?

File: docs/configuration.asciidoc [hunk 2]
```diff
@@ -2430,9 +2430,12 @@ Prepending an element with `(?-i)` makes the matching case sensitive.
...
+
 This configuration option helps to make MessageListener type matching faster and improve application startup performance.
 
+Starting from version 1.43.0, the classes that are part of the 'application_packages' option are also included.
```
Comment (jackshirazi) [inline docs/configuration.asciidoc:2437]: how does that affect the default vs non-default?

Comment (SylvainJuge) [reply docs/configuration.asciidoc:2437]: It does not affect it, it's just that classes that are part of the `application_packages` are also evaluated during class matching. In a sense it's just making the exclusion pre-filtering less restrictive than previously.

File: apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/configuration/MessagingConfiguration.java [hunk 3]
```diff
@@ -100,8 +100,11 @@ public class MessagingConfiguration extends ConfigurationOptionProvider {
         .configurationCategory(MESSAGING_CATEGORY)
         .description("Defines which packages contain JMS MessageListener implementations for instrumentation." +
             "\n" +
-            "When set to a non-empty value, only the classes matching configuration will be instrumented.\n" +
-            "This configuration option helps to make MessageListener type matching faster and improve application startup performance."
+            "When empty (default), only the inner-classes or that have 'Listener' or 'Message' in their names are considered.\n"+
```
Comment (jackshirazi) [inline apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/configuration/MessagingConfiguration.java:103]: only which inner classes? any in the whole application?

Comment (jackshirazi) [inline apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/configuration/MessagingConfiguration.java:103]:
```suggestion
            "When empty (default), all inner-classes or any classes that have 'Listener' or 'Message' in their names are considered.\n"+
```

File: docs/configuration.asciidoc (hunk 2 above)
Comment (jackshirazi) [reply docs/configuration.asciidoc:2437]: so the default option will be inner-classes, classes with 'Listener' or 'Message' in the name, and all classes in application_packages?

File: apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/configuration/MessagingConfiguration.java [hunk 4]
```diff
@@ -100,8 +100,11 @@ public class MessagingConfiguration extends ConfigurationOptionProvider {
         .configurationCategory(MESSAGING_CATEGORY)
         .description("Defines which packages contain JMS MessageListener implementations for instrumentation." +
             "\n" +
-            "When set to a non-empty value, only the classes matching configuration will be instrumented.\n" +
-            "This configuration option helps to make MessageListener type matching faster and improve application startup performance."
+            "When empty (default), all inner-classes or any classes that have 'Listener' or 'Message' in their names are considered.\n"+
+            "\n" +
+            "This configuration option helps to make MessageListener type matching faster and improve application startup performance.\n" +
+            "\n" +
+            "Starting from version 1.43.0, the classes that are part of the 'application_packages' option are also included."
```
Comment (jackshirazi) [inline apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/configuration/MessagingConfiguration.java:107]:
```suggestion
            "Starting from version 1.43.0, the classes that are part of the 'application_packages' option are also included in the list of classes considered."
```

**Review by jackshirazi — APPROVED** [review]

File: docs/configuration.asciidoc (hunk 2 above)
Comment (SylvainJuge) [reply docs/configuration.asciidoc:2437]: yes

**Pull request merged**